src/common/walFilter/walFilter.c

```diff
@@ -0,0 +1,763 @@
+        wrote += to_write;
+
+        bufUsedInc(output, to_write);
+
+        // write header
+        if (header_i < lstSize(this->headers))
+        {
+            to_write = SizeOfXLogShortPHD;
+            checkOutputSize(output, to_write);
+            memcpy(bufRemainsPtr(output), lstGet(this->headers, header_i), to_write);
+            bufUsedInc(output, to_write);
+            this->total_offset += to_write;
+            header_i++;
+        }
+    }
+
+    const size_t align_size = MAXALIGN(this->got_len) - this->got_len;
+    checkOutputSize(output, align_size);
+    memset(bufRemainsPtr(output), 0, align_size);
+    bufUsedInc(output, align_size);
+
+    this->total_offset += this->tot_len;
+    this->got_len = 0;
+}
+
+static bool
+readBeginOfRecord(WalFilterState *const main_state)
+{
+    bool result = false;
+    MEM_CONTEXT_TEMP_BEGIN();
+
+    const String *walSegment = NULL;
+    const TimeLineID timeLine = main_state->current_header->xlp_tli;
+    uint64_t segno = main_state->current_header->xlp_pageaddr / main_state->segSize;
+    const String *const path = strNewFmt("%08X%08X", timeLine, (uint32) (segno / XLogSegmentsPerXLogId(main_state->segSize)));
+
+    const StringList *const segmentList = storageListP(
+        storageRepoIdx(main_state->archiveInfo->repoIdx),
+        strNewFmt(STORAGE_REPO_ARCHIVE "/%s/%s", strZ(main_state->archiveInfo->archiveId), strZ(path)),
+        .expression = strNewFmt("^[0-f]{24}-[0-f]{40}" COMPRESS_TYPE_REGEXP "{0,1}$"));
+
+    if (strLstEmpty(segmentList))
+    {
+        // an exotic case where we couldn't even find the current file.
+        THROW(FormatError, "no WAL files were found in the repository");
+    }
+
+    uint64_t segno_diff = UINT64_MAX;
+    for (uint32_t i = 0; i < strLstSize(segmentList); i++)
+    {
+        const String *const file = strSubN(strLstGet(segmentList, i), 0, 24);
+        TimeLineID tli;
+        uint64_t fileSegNo = 0;
+        XLogFromFileName(strZ(file), &tli, &fileSegNo, main_state->segSize);
+
+        if (segno - fileSegNo < segno_diff && fileSegNo < segno)
+        {
+            segno_diff = segno - fileSegNo;
+            walSegment = strLstGet(segmentList, i);
+        }
+    }
+
+    // current file is oldest
+    if (segno_diff == UINT64_MAX)
+    {
+        memContextSwitchBack();
+        memContextDiscard();
+        return false;
+    }
+
+    const StorageRead *const storageRead = storageNewReadP(
+        storageRepoIdx(0),
+        strNewFmt(STORAGE_REPO_ARCHIVE "/%s/%s", strZ(main_state->archiveInfo->archiveId), strZ(walSegment)));
+
+    buildArchiveGetPipeLine(ioReadFilterGroup(storageReadIo(storageRead)), main_state->archiveInfo);
+
+    OBJ_NEW_BEGIN(WalFilterState, .childQty = MEM_CONTEXT_QTY_MAX, .allocQty = MEM_CONTEXT_QTY_MAX)
+    {
+        memset(this, 0, sizeof(WalFilterState));
+        this->is_begin = true;
+        this->record = memNew(this->heapPageSize);
```

Comment:
Isn't `heapPageSize` 0 here because of `memset` above?

Reply:
yes. it's should be `main_state->heapPageSize`